zebra-consensus/src/block/check.rs

```diff
@@ -101,14 +102,41 @@ pub fn subsidy_is_valid(block: &Block, network: Network) -> Result<(), BlockErro
     let height = block.coinbase_height().ok_or(SubsidyError::NoCoinbase)?;
     let coinbase = block.transactions.get(0).ok_or(SubsidyError::NoCoinbase)?;
 
+    // Consensus rule: The total value in zatoshi of transparent outputs from a
+    // coinbase transaction, minus vbalanceSapling, minus vbalanceOrchard, MUST NOT
+    // be greater than the value in zatoshi of block subsidy plus the transaction fees
+    // paid by transactions in this block.
+    // https://zips.z.cash/protocol/protocol.pdf#txnencodingandconsensus
+    let transparent_value_balance: Amount = subsidy::general::output_amounts(coinbase)
+        .iter()
+        .sum::<Result<Amount<NonNegative>, AmountError>>()
+        .expect("the sum of all outputs wil lalways be positive")
+        .constrain()
+        .expect("positive value always fit in `NegativeAllowed`");
+    let sapling_value_balance = coinbase.sapling_value_balance().sapling_amount();
+    let orchard_value_balance = coinbase.orchard_value_balance().orchard_amount();
+
+    let block_subsidy: Amount = subsidy::general::block_subsidy(height, network)
+        .expect("a valid block subsidy for this height and network")
+        .constrain()
+        .expect("positive value always fit in `NegativeAllowed`");
+
+    // Consensus rule implementation: block fee must be at least zero.
+    let fees: Result<Amount<NonNegative>, AmountError> =
+        (transparent_value_balance - sapling_value_balance - orchard_value_balance - block_subsidy)
+            .expect("should not overflow")
+            .constrain();
+
+    if fees.is_err() {
+        Err(SubsidyError::NegativeFees)?
+    }
```

Comment:
Idea: a suggestion for an alternative. Not sure if it makes things clearer, but it does join the code into a single block under the consensus rule comment :thinking: 

```suggestion
    (transparent_value_balance - sapling_value_balance - orchard_value_balance - block_subsidy)
        .expect("should not overflow")
        .constrain::<NonNegative>()
        .map_err(|_| SubsidyError::NegativeFees)?;
```

Reply:
I think this is better than what we had do i applied at https://github.com/ZcashFoundation/zebra/pull/3067/commits/19f66779d9b24b8b13ec044aec48881e5f660023